File: src/main/java/me/jsinco/solutilities/bukkitcommands/CommandMapper.java
package me.jsinco.solutilities.bukkitcommands;

import org.bukkit.Bukkit;
import org.bukkit.command.CommandMap;
import org.bukkit.command.defaults.BukkitCommand;

import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.Map;

public class CommandMapper {

    private static final Map<String, BukkitCommand> mappedCommands = new HashMap<>();

    /**
     * Registers a BukkitCommand to the server's command map using reflection, no need to declare in Plugin.yml
     * @param commandName The name of the command to register.
     * @param bukkitCommand The BukkitCommand to register.
     */

    // Paper gives access to the command map without NMS or reflection so...
    // I'm still going to leave it like this though
    // Guess it's still useful for Spigot projects though!
    public static void registerBukkitCommand(String commandName, BukkitCommand bukkitCommand) {
        try {
            final Field bukkitCommandMap = Bukkit.getServer().getClass().getDeclaredField("commandMap");

            bukkitCommandMap.setAccessible(true);
            CommandMap commandMap = (CommandMap) bukkitCommandMap.get(Bukkit.getServer());


            commandMap.register(commandName, "solutilities", bukkitCommand);
            mappedCommands.put(commandName, bukkitCommand);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            e.printStackTrace();
        }
    }

    public static void unRegisterBukkitCommands() {
        try {
            final Field bukkitCommandMap = Bukkit.getServer().getClass().getDeclaredField("commandMap");

            bukkitCommandMap.setAccessible(true);
            CommandMap commandMap = (CommandMap) bukkitCommandMap.get(Bukkit.getServer());


            for (BukkitCommand bukkitCommand : mappedCommands.values()) {
                commandMap.getCommand(bukkitCommand.getName()).unregister(commandMap);
                Bukkit.getCommandMap().getKnownCommands().remove(bukkitCommand.getName());
                for (String alias : bukkitCommand.getAliases()) {
                    Bukkit.getCommandMap().getKnownCommands().remove(alias);
                }
                Bukkit.getLogger().log(java.util.logging.Level.INFO, "Unregistered command: " + bukkitCommand.getName());
            }
        } catch (NoSuchFieldException | IllegalAccessException e) {
            e.printStackTrace();
        }
    }

}

File: src/main/java/me/jsinco/solutilities/bukkitcommands/CommandMapper.kt
package me.jsinco.solutilities.bukkitcommands

import me.jsinco.solutilities.SolUtilities
import me.jsinco.solutilities.bukkitcommands.commands.PingCommand
import me.jsinco.solutilities.bukkitcommands.commands.RanksMenu
import me.jsinco.solutilities.bukkitcommands.commands.SearchShopAlias
import org.bukkit.Bukkit
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandMap
import org.bukkit.command.PluginCommand
import java.lang.reflect.Field

class CommandMapper(val plugin: SolUtilities) {


    companion object {
        private val commands: MutableMap<String, CommandExecutor> = mutableMapOf()

        @JvmStatic
        fun register(name: String, commandExecutor: CommandExecutor) {
            commands[name] = commandExecutor
        }
    }


    init {
        // Commands incase they are not registered with method above
        commands["ls"] = SearchShopAlias(plugin)
        commands["lb"] = SearchShopAlias(plugin)
        commands["ranks"] = RanksMenu(plugin)
        commands["ping"] = PingCommand(plugin)

        // Mapper
        mapCommands()
    }

    private fun mapCommands() {
        try {
            val bukkitCommandMap: Field = Bukkit.getServer().javaClass.getDeclaredField("commandMap")
            bukkitCommandMap.isAccessible = true
            val commandMap: CommandMap = bukkitCommandMap.get(Bukkit.getServer()) as CommandMap

            for (command in commands) {
                val cmd = command.value

                val pluginCommand: PluginCommand = plugin.getCommand(command.key)!!
                pluginCommand.setExecutor(cmd)
                commandMap.register(command.key, "solutilities", pluginCommand)
            }
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
    }
}